src/host/posix/dnssd.cpp

```diff
@@ -672,21 +608,144 @@ void DnssdPlatform::StopAddressResolver(const AddressResolver
+            mTxtResolversMap.find(*iter) == mTxtResolversMap.end())
         {
-            entryList.CleanUpDeletedEntries();
-            if (entryList.IsEmpty())
-            {
-                aResolversMap.erase(iter);
-            }
+            mPublisher.UnsubscribeService(iter->GetType(), iter->GetInstance());
+            iter = mServiceNameSubscriptions.erase(iter);
         }
+        else
+        {
+            iter++;
+        }
+    }
+
+    // Subscribe to services (DnsServiceName) that haven't been subscribed.
+    for (const auto &entry : mServiceResolversMap)
+    {
+        const DnsServiceName &serviceName = entry.first;
+
+        if (mServiceNameSubscriptions.find(serviceName) == mServiceNameSubscriptions.end())
+        {
+            mServiceNameSubscriptions.insert(serviceName);
+            mPublisher.SubscribeService(serviceName.GetType(), serviceName.GetInstance());
+        }
+    }
+    for (const auto &entry : mTxtResolversMap)
+    {
+        const DnsServiceName &serviceName = entry.first;
+
+        if (mServiceNameSubscriptions.find(serviceName) == mServiceNameSubscriptions.end())
+        {
+            mServiceNameSubscriptions.insert(serviceName);
+            mPublisher.SubscribeService(serviceName.GetType(), serviceName.GetInstance());
+        }
+    }
+}
+
+void DnssdPlatform::PostServiceSubcriptionUpdateTask(void)
+{
+    if (!mServiceSubscriptionUpdateTaskPosted)
+    {
+        mTaskRunner.Post([this](void) { this->ExecuteServiceSubscriptionUpdate(); });
+        mHostSubscriptionUpdateTaskPosted = true;
+    }
+}
+
+void DnssdPlatform::ExecuteHostSubscriptionUpdate(void)
+{
+    mHostSubscriptionUpdateTaskPosted = false;
+
+    // Unsubscribe hosts (DnsName) that are stale.
+    for (auto iter = mHostSubscriptions.begin(); iter != mHostSubscriptions.end();)
+    {
+        const DnsName &dnsName = *iter;
+
+        if (mIp6AddrResolversMap.find(dnsName) == mIp6AddrResolversMap.end() &&
+            mIp4AddrResolversMap.find(dnsName) == mIp4AddrResolversMap.end())
+        {
+            mPublisher.UnsubscribeHost(dnsName.GetName());
+            iter = mHostSubscriptions.erase(iter);
+        }
+        else
+        {
+            iter++;
+        }
+    }
+
+    // Subscribe to hosts (DnsName) that haven't been subscribed.
+    for (auto &entry : mIp6AddrResolversMap)
+    {
+        const DnsName &dnsName = entry.first;
+
+        if (mHostSubscriptions.find(dnsName) == mHostSubscriptions.end())
+        {
+            mHostSubscriptions.insert(dnsName);
+            mPublisher.SubscribeHost(dnsName.GetName());
+        }
+    }
+    for (auto &entry : mIp6AddrResolversMap)
+    {
+        const DnsName &dnsName = entry.first;
+
+        if (mHostSubscriptions.find(dnsName) == mHostSubscriptions.end())
+        {
+            mHostSubscriptions.insert(dnsName);
+            mPublisher.SubscribeHost(dnsName.GetName());
+        }
+    }
```

Comment:
Probably you can merge these two loops into one to make it more concise.

Reply:
I've merged the 2 maps into one. Currently there is no difference to resolve Ipv6/Ipv4 host. The underlying to call mPublisher are exactly the same.